grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy

```diff
@@ -23,47 +57,150 @@ class PrepaymentInvoiceService {
 
         order.invoiceableOrderItems.each { OrderItem orderItem ->
             if (orderItem.orderItemStatusCode == OrderItemStatusCode.CANCELED) {
-                InvoiceItem invoiceItem = invoiceService.createFromOrderItem(orderItem)
+                InvoiceItem invoiceItem = createFromOrderItem(orderItem)
+                InvoiceItem inverseItem = createInverseItemForOrderItem(orderItem)
                 invoice.addToInvoiceItems(invoiceItem)
-            } else {
-                orderItem?.invoiceableShipmentItems?.each { ShipmentItem shipmentItem ->
-                    InvoiceItem invoiceItem = invoiceService.createFromShipmentItem(shipmentItem)
-                    invoice.addToInvoiceItems(invoiceItem)
+                invoice.addToInvoiceItems(inverseItem)
+                return
+            }
+
+            orderItem?.invoiceableShipmentItems?.each { ShipmentItem shipmentItem ->
+                InvoiceItem invoiceItem = createFromShipmentItem(shipmentItem)
+                InvoiceItem inverseItem = createInverseItemForShipmentItem(shipmentItem, invoiceItem)
+                invoice.addToInvoiceItems(invoiceItem)
+                if (inverseItem) {
+                    invoice.addToInvoiceItems(inverseItem)
                 }
             }
         }
 
         order.invoiceableAdjustments.each { OrderAdjustment orderAdjustment ->
-            InvoiceItem invoiceItem = invoiceService.createFromOrderAdjustment(orderAdjustment)
+            InvoiceItem invoiceItem = createFromOrderAdjustment(orderAdjustment)
+            InvoiceItem inverseItem = createInverseItemForOrderAdjustment(orderAdjustment)
             invoice.addToInvoiceItems(invoiceItem)
+            invoice.addToInvoiceItems(inverseItem)
         }
 
         return invoice.save()
     }
 
-    Invoice generatePrepaymentInvoice(Order order) {
-        if (order.orderItems.any { it.hasInvoices } || order.orderAdjustments.any { it.hasInvoices }) {
-            throw new Exception("Some order items or order adjustments for this order already have been invoiced")
+    InvoiceItem createFromOrderItem(OrderItem orderItem) {
```

Comment:
suggestion: make all these createFromX methods private. The only thing that needs to be public is whatever we're exposing to external sources (the controllers). That way it's clearer what the entrypoints to the class are